apps/arweave/src/ar_p3.erl

```diff
@@ -0,0 +1,325 @@
+%%% Public interface.
+%%%===================================================================
+allow_request(Req) ->
+	gen_server:call(?MODULE, {request, Req}).
+
+start_link() ->
+	gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+get_balance(Address, Network, Token) ->
+	gen_server:call(?MODULE, {get_balance, Address, ?TO_P3_ASSET(Network, Token)}).
+
+get_rates_json() ->
+	gen_server:call(?MODULE, {get_rates_json}).
+
+%%%===================================================================
+%%% Generic server callbacks.
+%%%===================================================================
+init([]) ->
+	process_flag(trap_exit, true),
+	ok = ar_events:subscribe(node_state),
+	{ok, Config} = application:get_env(arweave, config),
+	ar_p3_config:validate_config(Config).
+
+handle_call({request, Req}, _From, State) ->
+	case handle_request(Req, State) of
+		{ok, _} ->
+			{reply, {true, ok}, State};
+		{error, insufficient_funds} ->
+			{reply, {false, insufficient_funds}, State};
+		{error, stale_mod_seq} ->
+			{reply, {false, stale_mod_seq}, State};
+		{error, _Error} ->
+			{reply, {false, invalid_header}, State}
+	end;
+
+handle_call({get_balance, Address, Asset}, _From, State) ->
+	{reply, ar_p3_db:get_balance(Address, Asset), State};
+
+handle_call({get_rates_json}, _From, State) ->
+	{reply, ar_p3_config:get_json(State), State}.
+
+handle_cast(stop, State) ->
+	{stop, normal, State}.
+
+handle_info({event, node_state, {new_tip, B, _PrevB}}, State) ->
+	NumConfirmations = ar_p3_config:get_payments_value(
+							State, ?ARWEAVE_AR, #p3_payment.confirmations),
+	DepositAddress = ar_p3_config:get_payments_value(
+							State, ?ARWEAVE_AR, #p3_payment.address),
+	case {NumConfirmations, DepositAddress} of
+		{undefined, _} -> ok;
+		{_, undefined} -> ok;
+		{NumConfirmations, DepositAddress} ->
+			scan_blocks_for_deposits(B#block.height - NumConfirmations + 1, DepositAddress)
+	end,
+	{noreply, State};
+
+handle_info({event, node_state, _}, State) ->
+	{noreply, State}.
+
+terminate(Reason, State) ->
+	ok.
+
+%%%===================================================================
+%%% Private functions.
+%%%===================================================================
+
+%%--------------------------------------------------------------------
+%% handle_call({request}) helpers.
+%%
+%% Handle a web request and decide whether a charge is required and if
+%% so, levy that charge
+%%--------------------------------------------------------------------
+handle_request(Req, P3Config) when
+		is_record(P3Config, p3_config) ->
+	case ar_p3_config:get_service_config(P3Config, Req) of
+		undefined ->
+			{ok, not_p3_service};
+		ServiceConfig ->
+			validate_request(Req, P3Config, ServiceConfig)
+	end.
+
+validate_request(Req, P3Config, ServiceConfig) when
+		is_record(ServiceConfig, p3_service) ->
+	case validate_header(Req, ServiceConfig) of
+		{ok, Account} ->
+			apply_charge(Account, Req, P3Config, ServiceConfig);
```

Comment:
Good point- it shouldn't be that hard to change. I think we'll need to validate the request twice though (first time is a full header/signature/balance validation, second time just checks the account balance), like:

1. Request comes in
2. validate request (by calling `allow_request`)
3. handle the request
4. validate the balance
5. apply the charge
6. issue the response

If we don't validate twice we risk overspending if a client issues a lot of requests at the same time.

Reply:
Ha, I implemented the above, but now I think we might have to do it this way:
1. Request comes in
2. Validate the request *and* apply the charge
3. handle the request
4. If the request fails: *reverse* the charge
5. issue the response

Otherwise there are still a couple spam attack vectors (where a malicious client issues many concurrent requests)
1. For expensive GET requests a client could cause a node to spend a lot of time building the GET response before realizing that the client does not have sufficient balance to pay for all the requests.
2. For POST requests the node would end up updating state for each concurrent request before realizing, at the end, that only a subset of them can be paid for.

I think the charge->process->reverse should negate the spam attacks. In both cases the spammy requests would be blocked before they were processed - and in the case wher there's a non-malicious request failure, the charge would be reversed.

Reply:
Done! Implemented in these commits: https://github.com/ArweaveTeam/arweave/compare/75113e1..dff1e3d